harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -920,6 +927,501 @@ def mock_get_tensorboard_path(dummy: typing.Dict[str, typing.Any]) -> pathlib.Pa
+            nproc_per_node=2,
+            run_id=rdzv_id,
+            max_restarts=0,
+            rdzv_endpoint=rdzv_endpoint,
+            rdzv_backend=rdzv_backend,
+        )
+
+        outputs = launcher.elastic_launch(launch_config, self.run_amp)(tmp_path, api_style)
+        outputs = launcher.elastic_launch(launch_config, self.run_amp)(
+            tmp_path, api_style, outputs[0]
+        )
+
+    @pytest.mark.skipif(torch.cuda.device_count() < 2, reason="not enough gpus")
+    @pytest.mark.gpu_parallel
+    def test_distributed_logging(self, tmp_path: pathlib.Path):
+        # set up distributed backend.
+        os.environ[det._DistributedBackend.TORCH] = str(1)
+
+        rdzv_backend = "c10d"
+        rdzv_endpoint = "localhost:29400"
+        rdzv_id = str(uuid.uuid4())
+
+        num_procs = 2
+
+        launch_config = launcher.LaunchConfig(
+            min_nodes=1,
+            max_nodes=1,
+            nproc_per_node=num_procs,
+            run_id=rdzv_id,
+            max_restarts=0,
+            rdzv_endpoint=rdzv_endpoint,
+            rdzv_backend=rdzv_backend,
+        )
+
+        outputs = launcher.elastic_launch(launch_config, self.run_no_op)(tmp_path)
+
+        log_output = sum([outputs[i] for i in range(num_procs)], [])
+
+        patterns = [f"finished train_batch for rank {i}" for i in range(num_procs)]
+
+        utils.assert_patterns_in_logs(log_output, patterns)
+
+    @pytest.mark.skipif(torch.cuda.device_count() < 2, reason="not enough gpus")
+    @pytest.mark.gpu_parallel
+    @pytest.mark.parametrize("dataset_len", [2, 3])
+    def test_epoch_sync(self, tmp_path: pathlib.Path, dataset_len: int):
+        # set up distributed backend.
+        os.environ[det._DistributedBackend.TORCH] = str(1)
+
+        rdzv_backend = "c10d"
+        rdzv_endpoint = "localhost:29400"
+        rdzv_id = str(uuid.uuid4())
+
+        num_procs = 2
+
+        launch_config = launcher.LaunchConfig(
+            min_nodes=1,
+            max_nodes=1,
+            nproc_per_node=num_procs,
+            run_id=rdzv_id,
+            max_restarts=0,
+            rdzv_endpoint=rdzv_endpoint,
+            rdzv_backend=rdzv_backend,
+        )
+
+        num_steps = 10
+        global_batch_size = 2
+        outputs = launcher.elastic_launch(launch_config, self.run_no_op)(
+            tmp_path, num_steps, global_batch_size, dataset_len
+        )
+
+        log_output = sum([outputs[i] for i in range(num_procs)], [])
+
+        batches_per_epoch = (dataset_len + global_batch_size - 1) // global_batch_size  # ceil
+
+        patterns = []
+        for rank in range(num_procs):
+            for batch_idx in range(num_steps):
+                epoch_idx = batch_idx // batches_per_epoch
+                patterns.append(f"rank {rank} finished batch {batch_idx} in epoch {epoch_idx}")
+
+        utils.assert_patterns_in_logs(log_output, patterns)
+
+    def run_mnist(self, tmp_path: pathlib.Path, batches_trained: typing.Optional[int] = 0):
```

Comment:
I think using the test class for `self` isn't very necessary, and really inflates the test class.

It looks like it would be fairly easy to write `run_mnist`, `run_gan`, etc as standalone functions with no `self`.

It looks like their only use of `self` is for:
- `self.train_and_checkpoint`
- `self.train_for_checkpoint`
- `self.train_from_checkpoint`

And those functions, in turn, only depend on `self` for `self.trial_seed`, which is literally 17 in almost every single usage in this whole file.  That could be set as a default value to the `trial_seed` parameter of `create_trial_and_trial_controller()`.

So definitely break things up along those lines.

Reply:
You were talking about separating things out of this file; I think "remove things from this file which aren't themselves unit tests, and put them in a helper file in this directory" is a good idea.

Reply:
I refactored a chunk of functions into `pytorch_utils.py` and a new `pytorch/test_examples.py`